scripts/india_census/primary_religion_data/preprocess.py

```diff
@@ -0,0 +1,354 @@
+        self.dataset_name = dataset_name
+        self.raw_df = None
+        self.stat_var_index = {}
+        self.census_columns = []
+        self.data_categories = data_categories
+        self.data_category_column = data_category_column
+
+    def _format_data(self):
+
+        def stat_var_index_key(row, data_category_column=None):
+            key = "{0}_{1}".format(row["columnName"], row["TRU"])
+            if data_category_column:
+                key = key + "_" + row[data_category_column]
+            return key
+
+        #Generate Census locationid
+        self.raw_df["census_location_id"] = self.raw_df.apply(
+            self._format_location, axis=1)
+
+        #Remove the unwanted columns. They are census codes which we dont use
+        #State,District,Subdistt,Town/Village,Ward,EB
+        #We delete them only if they exists
+        #From pandas documentation:
+        #If errors=‘ignore’, suppress error and only existing labels are dropped
+        self.raw_df.drop(
+            ["State", "District", "Subdistt", "Town/Village", "Name"],
+            axis=1,
+            inplace=True,
+            errors='ignore')
+
+        #Once the above columns are dropped, we will have
+        #First column = TRU/placeOfResidence
+        #Second column = data_category_column
+        #3-N are the actual values. Let's get value columns
+        value_columns = list(self.raw_df.columns[2:-1])
+
+        #convert the data_category_column to titleCase
+        self.raw_df[self.data_category_column] = self.raw_df[
+            self.data_category_column].apply(lambda x: titleCase(x))
+
+        #converting rows in to columns. So the final structure will be
+        #Name,TRU,columnName,value
+        self.raw_df = self.raw_df.melt(
+            id_vars=["census_location_id", "TRU", self.data_category_column],
+            value_vars=value_columns,
+            var_name='columnName',
+            value_name='Value')
+
+        #Add corresponding StatisticalVariable, based on columnName and TRU
+        self.raw_df['StatisticalVariable'] = self.raw_df.apply(
+            lambda row: self._get_stat_var_name(self.stat_var_index[
+                stat_var_index_key(row, self.data_category_column)]),
+            axis=1)
+
+        #add the census year
+        self.raw_df['Year'] = self.census_year
+
+        #remove data_category_column as its not required
+        self.raw_df.drop([self.data_category_column],
+                         axis=1,
+                         inplace=True,
+                         errors='ignore')
+
+        #Export it as CSV. It will have the following columns
+        #Name,TRU,columnName,value,StatisticalVariable,Year
+        self.raw_df.to_csv(self.csv_file_path, index=False, header=True)
+
+    def _get_base_name(self, row):
+        #To make the name meanigful add Religion to the
+        #the name of the stat var
+        name = "Count_" + row["populationType"] + "_Religion"
```

Comment:
Is this same as self.data_category_column ?

Reply:
They have the same value. But `self.data_category_column` is the name of the column that has religions in the xls.